Add periodic corner neighbours

TomMelt opened the pull request:
## Add Periodic neighbours to metadata
 
This PR does the following:

- adds logic to identify periodic corner neighbours.
- adds functionality to write them to the metadata file. 
- Whilst I have updated the existing integration test to pass with the new output, I haven't explicitly tested periodic neighbours in the integration test.

## TODO

- [x] need to create arrays to store periodic neighbours
- [x] need to output these arrays in `Partitioner::save_metadata`
- [x] need to add tests (integration)

TomMelt committed:
wip: add periodic neighbour logic

* need to create arrays to store periodic neighbours
* need to output these arrays in Partitioner::save_metadata
* need to add tests (integration)

TomMelt committed:
chore: match naming convention of main branch

TomMelt committed:
chore: use std:array to match rest of code

TomMelt committed:
feat: output periodic corner in save_metadata

TomMelt committed:
test: update integration tests with new output

niravshah241 commented:
Changes overall look good to me. I will be adding some tests to see it running. I guess most changes are related to:

- Changing ``vector`` with ``array`` 
- Addition of periodic logic
- Change of variable name for readability and/or uniformity with other code

Is their something else that I missed?

Commit:
Periodic corner neighbour test case

Commit:
Periodic corner neighbour test case

Commit:
Periodic corner neighbour tests test_4_px test_4_py test_4_px_py added

Commit:
Branch unnecesary files cleanup

niravshah241 reviewed (approved):
Approved. One more test will be added (for top and bottom overlapping neighbour) in PR https://github.com/nextsimhub/domain_decomp/pull/76

TomMelt commented on Partitioner.cpp:
```diff
@@ -150,32 +155,57 @@ void Partitioner::haloBufferPositions(
 }
 
-int Partitioner::halo_corner_start(const Domain d1, const Domain d2, const Vertex vertex)
+int Partitioner::halo_corner_start(
```
@niravshah241 just flagging, I didn't verify/test `Partitioner::halo_corner_start()` for periodic neighbours. So it might be incorrect.

Commit:
Unit test for is_corner_neighbour and halo_corner_start added

Commit:
Unit test for is_corner_neighbour and halo_corner_start added

Commit:
Periodic unit tests added in test_halo_corner_start and test_is_corner_neighbour

TomMelt committed:
wip: initial fix for non-periodic neighbours

Commit:
test_halo_corner_start updated

Commit:
Partitioner class updated making haloBufferPositions, halo_corner_start, is_neighbour, is_corner_neighbour public members (which were private members earlier)

TomMelt committed:
feat: fix corner neighbour for periodic BC

TomMelt committed:
test: update test reference data

TomMelt committed:
wip: Nirav see if you can find why its breaking

Commit:
Tests updated and fixes in Partition.cpp

Commit:
DOmain decomp test in works

Commit:
Periodic tests added

TomMelt committed:
doc: correct typo in comment regarding haloBufferPositions

TomMelt committed:
feat: fix haloCornerBufferPositions logic

TomMelt committed:
test: generalize test_haloCornerBufferPositions.cpp

TomMelt committed:
test: update reference data for integration test

Commit:
test_haloBufferPositions updated

Commit:
domain decomp image

Commit:
is_neighbour test fixed, corrections in is_neighbour function

TomMelt committed:
chore: use std::vector instead of list

TomMelt committed:
chore: move edge neighbour check logic to is_neighbour

remove neighbour check logic from domain_overlap

update comments in is_neighbour to reflect the changes

in Partitioner we still check the halo_size is non-zero but this will
always be true, so it is not a problem.

TomMelt committed:
chore: update doc for mod function

TomMelt committed:
chore: update filenames in header

TomMelt committed:
test: remove redundant tests

TomMelt committed:
chore: remove unused vars

TomMelt committed:
chore: tidy buffer tests

TomMelt committed:
chore: rename haloBufferPositions to haloEdgeBufferPositions

TomMelt committed:
chore: rename vertices to corners

TomMelt committed:
chore: rename vars to camelCase to match nextsim

TomMelt committed:
chore: apply formatting changes

Co-authored-by: Joe Wallwork <22053413+joewallwork@users.noreply.github.com>
